src/NUnitFramework/framework/Constraints/ExactCountConstraint.cs

```diff
@@ -69,21 +70,32 @@ public ExactCountConstraint(int expectedCount, IConstraint itemConstraint)
         public override ConstraintResult ApplyTo<TActual>(TActual actual)
         {
             var enumerable = ConstraintUtils.RequireActual<IEnumerable>(actual, nameof(actual));
+            var itemList = new Collection<object>();
+            var matchCount = 0;
 
-            int count = 0;
             if (_itemConstraint == null)
             {
-                foreach (object item in enumerable)
-                    count++;
+                foreach (var item in enumerable)
+                {
+                    matchCount++;
+
+                    if (itemList.Count <= ExactCountConstraintResult.MaxDisplayCount)
+                        itemList.Add(item);
+                }
             }
             else
             {
-                foreach (object item in enumerable)
+                foreach (var item in enumerable)
+                {
                     if (_itemConstraint.ApplyTo(item).IsSuccess)
-                        count++;
+                        matchCount++;
+
+                    if (itemList.Count <= ExactCountConstraintResult.MaxDisplayCount)
+                        itemList.Add(item);
```

Comment:
I was confused for a while, so it might be good to add a comment saying that we intentionally add one item too many to trigger the ellipsis when we later on call `FormatCollection`.